ocaml/src/repair.ml

```diff
@@ -0,0 +1,216 @@
+              ~namespace_id
+              manifest
+              ~chunk_id ~version_id ~gc_epoch
+              locations
+              compression encryption
+              fragment_checksum_algo
+              ~is_replication:(k=1))
+           (fun () ->
+            List.iter
+              Lwt_bytes.unsafe_destroy
+              all_fragments;
+            Lwt.return ())
+         >>= fun updated_locations ->
+
+         Lwt.return (chunk_id, updated_locations)
+       end)
+    (List.mapi (fun i lc -> i, lc) fragment_info)
+  >>= fun updated_locations ->
+
+  let updated_object_locations =
+    List.fold_left
+      (fun acc (chunk_id, updated_locations) ->
+       let updated_chunk_locations =
+         List.map
+           (fun (fragment_id, device_id) ->
+            (chunk_id, fragment_id, device_id))
+           updated_locations in
+       List.rev_append updated_chunk_locations acc)
+      []
+      updated_locations
+  in
+
+  Lwt.return (updated_object_locations, gc_epoch, version_id)
+
+let repair_object_generic_and_update_manifest
+      alba_client
+      ~namespace_id
+      ~manifest
+      ~problem_osds
+      ~problem_fragments =
+  repair_object_generic
+    alba_client
+    ~namespace_id
+    ~manifest
+    ~problem_osds
+    ~problem_fragments
+  >>= fun (updated_object_locations, gc_epoch, version_id) ->
+
+  let open Nsm_model in
+  let object_name = manifest.Manifest.name in
+  let object_id = manifest.Manifest.object_id in
+
+  alba_client # with_nsm_client'
+              ~namespace_id
+              (fun client ->
+               client # update_manifest
+                      ~object_name
+                      ~object_id
+                      (List.map
+                         (fun (c,f,o) -> c,f,Some o)
+                         updated_object_locations)
+                      ~gc_epoch ~version_id)
+  >>= fun () ->
+  Lwt_log.debug_f
+    "updated_manifest ~namespace_id:%li ~object_id:%S ~updated_object_locations:%s"
+    namespace_id object_id
+    ([%show : (int *int * int32) list] updated_object_locations)
+  >>= fun () ->
+  Lwt.return ()
+
+let rewrite_object
+      alba_client
+      ~namespace_id
+      ~manifest =
+  let open Nsm_model in
+
+  Lwt_log.debug_f
+    "Repairing %s in namespace %li"
+    (Manifest.show manifest) namespace_id
+  >>= fun () ->
+
+  let object_reader =
+    new Object_reader2.object_reader
+        alba_client
+        namespace_id manifest in
+
+  let object_name = manifest.Manifest.name in
+  let checksum_o = Some manifest.Manifest.checksum in
+  let allow_overwrite = PreviousObjectId manifest.Manifest.object_id in
+
+  Lwt.catch
+    (fun () ->
+     alba_client # upload_object'
+                 ~namespace_id
+                 ~object_name
+                 ~object_reader
+                 ~checksum_o
+                 ~allow_overwrite >>= fun _ ->
+     Lwt.return ())
+    (let open Nsm_model.Err in
+     function
+     | Nsm_exn (Overwrite_not_allowed, _) ->
```

Comment:
Pushed this error handling down into rewrite_object so not all it's callers have to do this